### DISCUSSION TITLE
Custom block jittery behavior

### DISCUSSION THREAD

**Discussion by kohn12** [OpenModelica/OpenModelica#11965] [Q&A] [answered]
In my OpenModelica model, I added a custom block to simulate the behavior of a push-pull bar. This behavior can be seen as three different spring stiffnesses. Which spring stiffness needs to be used is based on the position of the push-pull bar. I modelled this with an if-statement as can be seen below:

    block VariableSpringStiffness
      extends Modelica.Mechanics.Translational.Interfaces.PartialCompliant;
      import Modelica.Units.SI;
    
      // Variable instead of parameter
      SI.TranslationalSpringConstant c "Spring constant";
      parameter SI.Distance s_rel0=0 "Unstretched spring length";
      Real s1 = 1/1000 "Slack M90 bolt [m]";
      Real s2 = 12/1000 "Displacement Push-Pull bar [m]";
      Real f0 = 281e3;
      
    equation
      // Initialize variable c based on conditions
      if abs(s_rel) < s1 then
        c = f0/s1 "Preload over slack [N/m]";
      elseif abs(s_rel) < s2 then
        c = 35e3/s2 "Spring stiffness";
      else
        c = 2e11 "Material stiffness";
      end if;

      // Rest of the equations
      f = 2*c * (s_rel - s_rel0);  // 2 springs
    end VariableSpringStiffness;

This code works, but the behavior is very jittery, as the spring stiffness sometimes changes per time step. Is there a way to smooth this behavior out, preferably without interpolation and using the spring stiffness provided.

**Comment by casella** [comment]
@kohn12 you should change the _differential_ stiffness, not the _total_ stiffness. The way you are doing this, as soon as you cross the thresholds, the force f changes discontinuously, and this of course leads to jittery behaviour. What you need is a piecewise-linear continuous force-displacement law. You can use the Modelica.Blocks.Tables.CombiTable1Ds for that.

**Reply by kohn12** [reply]
Thanks for the answer! A question i have about this is if it will still work in a mechanical model. Currently the custom spring I added is placed over a prismatic joint, would this be possible to do with a CombiTable1Ds? 
Should I use a sensor to get the position, use this as input for the CombiTable to find the spring stiffness, and then input the spring stiffness into the code for a spring?

**Reply by casella** [reply, marked as answer]
I would create a custom 1D translational nonlinear spring by inheriting from Modelica.Mechanics.Translational.Interfaces.PartialCompliant, using CombiTimeTable1D to represent the piecewise-linear force-displacement law, then connect it to the two 1D connectors of the prismatic joint you can get by setting useAxisFlange = true. 

Using sensors is a bit ugly from an object-oriented point of view, they should only be used to get signals out to control systems

**Reply by kohn12** [reply]
I changed my model with your tips, it seems to work now. I added the code below, does it seem logical to you? Thanks!
```modelica
    block VariableSpringStiffness
       extends Modelica.Mechanics.Translational.Interfaces.PartialCompliant;
       import Modelica.Blocks.Types.Smoothness;
       import Modelica.Units.SI;
    
      // Define input
      input Real s_rel "Relative displacement";

      // Define output
       output Real f "Force";
       output SI.TranslationalSpringConstant c "Spring constant";

      // Define variables for the spring parameters
      parameter Real s1 = 1/1000 "Slack M90 bolt [m]";
      parameter Real s2 = 12/1000 "Displacement Push-Pull bar [m]";
      parameter Real f0 = 281e3 "Preload"; 
    
      // Define CombiTable1Ds for determining spring constant c
       Modelica.Blocks.Tables.CombiTable1Ds springTable(
        tableOnFile=false,
        tableName="SpringTable",
        columns={2},
        fileName="",
        table = [0, f0/s1; s1, f0/s1;s1+1e-5,35e3/s2; s2, 35e3/s2; s2+1e-5, 2e11;s2+1,2e11]); // Data points: s_rel, c

      // Define intermediate variable for absolute value of s_rel
      Real abs_s_rel;
    
    equation
      // Calculate absolute value of s_rel
      abs_s_rel = abs(s_rel);
    
      // Connect input to CombiTable1Ds
      springTable.u = abs_s_rel;
    
      // Assign spring constant c from CombiTable1Ds
      c = springTable.y[1];
    
      // Calculate force
      f = 2*c * s_rel; // 2 springs used
    end VariableSpringStiffness;
```

**Reply by casella** [reply]
No, sorry, that's not what I meant. My suggestion was to use the CombiTable to repesent the relationship between relative displacement and _force_, not between relative displacement and _spring constant_. 

The spring constant concept is useful if it is, well, constant. Otherwise, forget about it 😄

**Reply by kohn12** [reply]
That makes sense, thank you!
